readme.rst
Oridomi
========

django CMS Oridomi integration


HISTORY
CHANGELOG
=========

0.0.1 (2018-01-17)
------------------

* Initial project setup


ripiu/cmsplugin_oridomi/models.py
from django.db import models
from django.core.validators import (
    MaxValueValidator, MinValueValidator, validate_comma_separated_integer_list
)
from django.utils.translation import ugettext_lazy as _

from cms.models import CMSPlugin


class OridomiPlugin(CMSPlugin):
    '''Oridomi container'''

    SHADING_HARD = 'hard'
    SHADING_SOFT = 'soft'
    SHADING_DISABLED = 'false'
    SHADING_CHOICES = (
        (SHADING_DISABLED, _('disabled')),
        (SHADING_HARD, _('hard')),
        (SHADING_SOFT, _('soft')),
    )

    RIPPLE_DISABLED = 0
    RIPPLE_FORWARD = 1
    RIPPLE_BACKWARDS = 2
    RIPPLE_CHOICES = (
        (RIPPLE_DISABLED, _('disabled')),
        (RIPPLE_FORWARD, _('forward')),
        (RIPPLE_BACKWARDS, _('backwards')),
    )

    vertical_panels = models.CharField(
        _('vertical panels'), max_length=255,
        validators=[validate_comma_separated_integer_list],
        blank=False, default='3',
        help_text=_('The number of vertical panels (for folding left or '
                    'right). You can use either an integer, or an array of '
                    'percentages if you want custom panel widths, e.g. 20, '
                    '10, 10, 20, 10, 20, 10. The numbers must add up to 100 '
                    '(or near it, so you can use values like 33, 33, 33).')
    )

    horizontal_panels = models.CharField(
        _('horizontal panels'), max_length=255,
        validators=[validate_comma_separated_integer_list],
        blank=False, default='3',
        help_text=_('The number of horizontal panels (for folding top or '
                    'bottom) or an array of percentages.')
    )

    perspective = models.PositiveSmallIntegerField(
        _('perspective'),
        blank=False, null=False, default=1000,
        help_text=_('The determines the distance in pixels (z axis) of the '
                    'camera/viewer to the paper. The smaller the value, the '
                    'more distorted and exaggerated the effects will appear.')
    )

    shading = models.CharField(
        _('shading'), max_length=255,
        choices=SHADING_CHOICES, default=SHADING_HARD,
        help_text=_("The default shading style is hard, which shows distinct "
                    "creases in the paper. Other options include 'soft' — for "
                    "a smoother, more rounded look — or false to disable "
                    "shading altogether for a flat look.")
    )

    speed = models.PositiveSmallIntegerField(
        _('speed'),
        blank=False, null=False, default=700,
        help_text=_('Determines the duration of all animations in '
                    'milliseconds.')
    )

    max_angle = models.PositiveSmallIntegerField(
        _('maximum angle'),
        blank=False, null=False, default=90,
        help_text=_('Configurable maximum angle for effects. With most '
                    'effects, exceeding 90/-90 usually makes the element wrap '
                    'around and pass through itself leading to some glitchy '
                    'visuals.')
    )

    ripple = models.PositiveSmallIntegerField(
        _('ripple'),
        blank=False, null=False,
        choices=RIPPLE_CHOICES, default=RIPPLE_DISABLED,
        help_text=_('Ripple mode causes effects to fold in a staggered, '
                    'cascading manner. 1 indicates a forward cascade, 2 is '
                    'backwards. It is disabled by default.')
    )

    oridomi_class = models.CharField(
        _('Oridomi class'), max_length=100,
        blank=True, default='',
        help_text=_('This CSS class is applied to OriDomi elements so they '
                    'can be easily targeted later.')
    )

    shading_intensity = models.DecimalField(
        _('shading intensity'),
        max_digits=3, decimal_places=2,
        blank=False, null=False, default=1,
        validators=[MinValueValidator(0), MaxValueValidator(1)],
        help_text=_('This is a multiplier that determines the darkness of '
                    'shading. If you need subtler shading, set this to a '
                    'value below 1.')
    )

    # TODO: write a validator for this
    easing_method = models.CharField(
        _('easing method'), max_length=255,
        blank=True, default='',
        help_text=_('This option allows you to supply the name of a CSS '
                    'easing method or a cubic bezier formula for customized '
                    'animation easing.')
    )

    gap_nudge = models.DecimalField(
        _('gap nudge'),
        max_digits=3, decimal_places=1,
        blank=False, null=False, default=1.5,
        validators=[MinValueValidator(0)],
        help_text=_('Number of pixels to offset each panel to prevent small '
                    'gaps from appearing between them. This is configurable '
                    'if you have a need for precision.')
    )

    touch_enabled = models.BooleanField(
        _('touch enabled'), default=True,
        help_text=_('Allows the user to fold the element via touch or mouse.')
    )

    touch_sensitivity = models.DecimalField(
        _('touch sensitivity'),
        max_digits=4, decimal_places=2,
        blank=False, null=False, default=0.25,
        help_text=_('Coefficient of touch/drag action’s distance delta. '
                    'Higher numbers cause more movement.')
    )

    touch_start_callback = models.CharField(
        _('touchstart callback'),
        max_length=255, blank=True,
        help_text=_('Invoked with starting coordinate as first argument.')
    )

    touch_move_callback = models.CharField(
        _('touchmove callback'),
        max_length=255, blank=True,
        help_text=_('Invoked with the folded angle.')
    )

    touch_end_callback = models.CharField(
        _('touchend callback'),
        max_length=255, blank=True,
        help_text=_('Invoked with ending point.')
    )

    def __str__(self):
        return str(_('(%(columns)s, %(rows)s)' % {
            'columns': self.vertical_panels,
            'rows': self.horizontal_panels,
        }))

    class Meta:
        verbose_name = _('Oridomi container')
        verbose_name_plural = _('Oridomi containers')


ripiu/cmsplugin_oridomi/cms_plugins.py
from django.utils.html import mark_safe
from django.utils.translation import ugettext_lazy as _

from cms.plugin_base import CMSPluginBase
from cms.plugin_pool import plugin_pool

from . import __version__
from .models import OridomiPlugin

def panels(ints):
    result = [int(x) for x in ints.split(',')]
    if len(result) == 1:
        return result[0]
    return result

@plugin_pool.register_plugin
class OridomiPluginPublisher(CMSPluginBase):
    model = OridomiPlugin
    name = 'Oridomi'
    module = 'Ri+'
    render_template = 'ripiu/cmsplugin_oridomi/oridomi.html'
    allow_children = True
    fieldsets = (
        ('', {
            'fields': (
                ('vertical_panels', 'horizontal_panels'),
                'perspective',
                'shading',
                'speed',
                'max_angle',
                'ripple',
                'oridomi_class',
                'shading_intensity',
                'easing_method',
                'gap_nudge',
                'touch_enabled',
                'touch_sensitivity',
            ),
        }), (
            _('Callbacks'), {
                'description': _('Custom callbacks for touch/drag events. '
                                 'Each one is invoked with a relevant value '
                                 'so they can be used to manipulate objects '
                                 'outside of the OriDomi instance (e.g. '
                                 'sliding panels). x values are returned '
                                 'when folding left and right, y values for '
                                 'top and bottom. The second argument passed '
                                 'is the original touch or mouse event. '
                                 'These are empty functions by default.'),
                'fields': (
                    'touch_start_callback',
                    'touch_move_callback',
                    'touch_end_callback',
                ),
            }
        )
    )

    def render(self, context, instance, placeholder):
        import json
        context = super(OridomiPluginPublisher, self).render(
            context, instance, placeholder
        )
        oridomi_conf = {
            'vPanels': panels(instance.vertical_panels),
            'hPanels': panels(instance.horizontal_panels),
            'perspective': instance.perspective,
            'shading': instance.shading,
            'speed': instance.speed,
            'maxAngle': instance.max_angle,
            'ripple': instance.ripple,
            
            # Object of type 'Decimal' is not JSON serializable
            'shadingIntensity': str(instance.shading_intensity),
            
            'easingMethod': instance.easing_method,
            
            # Object of type 'Decimal' is not JSON serializable
            'gapNudge': str(instance.gap_nudge),
            
            'touchEnabled': instance.touch_enabled,
            
            # Object of type 'Decimal' is not JSON serializable
            'touchSensitivity': str(instance.touch_sensitivity),
        }
        if instance.oridomi_class:
            oridomi_conf['oriDomiClass'] = instance.oridomi_class
        if instance.touch_start_callback:
            oridomi_conf['touchStartCallback'] = instance.touch_start_callback
        if instance.touch_move_callback:
            oridomi_conf['touchMoveCallback'] = instance.touch_move_callback
        if instance.touch_end_callback:
            oridomi_conf['touchEndCallback'] = instance.touch_end_callback
        context.update({
            'instance': instance,
            'placeholder': placeholder,
            'version': __version__,
            'conf': mark_safe(json.dumps(oridomi_conf)),
        })
        return context


ripiu/cmsplugin_oridomi/migrations/0001_initial.py
# -*- coding: utf-8 -*-
# Generated by Django 1.10.8 on 2018-01-17 18:25
from __future__ import unicode_literals

import django.core.validators
from django.db import migrations, models
import django.db.models.deletion
import re


class Migration(migrations.Migration):

    initial = True

    dependencies = [
        ('cms', '0016_auto_20160608_1535'),
    ]

    operations = [
        migrations.CreateModel(
            name='OridomiPlugin',
            fields=[
                ('cmsplugin_ptr', models.OneToOneField(auto_created=True, on_delete=django.db.models.deletion.CASCADE, parent_link=True, primary_key=True, related_name='cmsplugin_oridomi_oridomiplugin', serialize=False, to='cms.CMSPlugin')),
                ('vertical_panels', models.CharField(default='3', help_text='The number of vertical panels (for folding left or right). You can use either an integer, or an array of percentages if you want custom panel widths, e.g. 20, 10, 10, 20, 10, 20, 10. The numbers must add up to 100 (or near it, so you can use values like 33, 33, 33).', max_length=255, validators=[django.core.validators.RegexValidator(re.compile('^\\d+(?:\\,\\d+)*\\Z', 32), code='invalid', message='Enter only digits separated by commas.')], verbose_name='vertical panels')),
                ('horizontal_panels', models.CharField(default='3', help_text='The number of horizontal panels (for folding top or bottom) or an array of percentages.', max_length=255, validators=[django.core.validators.RegexValidator(re.compile('^\\d+(?:\\,\\d+)*\\Z', 32), code='invalid', message='Enter only digits separated by commas.')], verbose_name='horizontal panels')),
                ('perspective', models.PositiveSmallIntegerField(default=1000, help_text='The determines the distance in pixels (z axis) of the camera/viewer to the paper. The smaller the value, the more distorted and exaggerated the effects will appear.', verbose_name='perspective')),
                ('shading', models.CharField(choices=[('false', 'disabled'), ('hard', 'hard'), ('soft', 'soft')], default='hard', help_text="The default shading style is hard, which shows distinct creases in the paper. Other options include 'soft' — for a smoother, more rounded look — or false to disable shading altogether for a flat look.", max_length=255, verbose_name='shading')),
                ('speed', models.PositiveSmallIntegerField(default=700, help_text='Determines the duration of all animations in milliseconds.', verbose_name='speed')),
                ('max_angle', models.PositiveSmallIntegerField(default=90, help_text='Configurable maximum angle for effects. With most effects, exceeding 90/-90 usually makes the element wrap around and pass through itself leading to some glitchy visuals.', verbose_name='maximum angle')),
                ('ripple', models.PositiveSmallIntegerField(choices=[(0, 'disabled'), (1, 'forward'), (2, 'backwards')], default=0, help_text='Ripple mode causes effects to fold in a staggered, cascading manner. 1 indicates a forward cascade, 2 is backwards. It is disabled by default.', verbose_name='ripple')),
                ('oridomi_class', models.CharField(blank=True, default='', help_text='This CSS class is applied to OriDomi elements so they can be easily targeted later.', max_length=100, verbose_name='Oridomi class')),
                ('shading_intensity', models.DecimalField(decimal_places=2, default=1, help_text='This is a multiplier that determines the darkness of shading. If you need subtler shading, set this to a value below 1.', max_digits=3, validators=[django.core.validators.MinValueValidator(0), django.core.validators.MaxValueValidator(1)], verbose_name='shading intensity')),
                ('easing_method', models.CharField(blank=True, default='', help_text='This option allows you to supply the name of a CSS easing method or a cubic bezier formula for customized animation easing.', max_length=255, verbose_name='easing method')),
                ('gap_nudge', models.DecimalField(decimal_places=1, default=1.5, help_text='Number of pixels to offset each panel to prevent small gaps from appearing between them. This is configurable if you have a need for precision.', max_digits=3, validators=[django.core.validators.MinValueValidator(0)], verbose_name='gap nudge')),
                ('touch_enabled', models.BooleanField(default=True, help_text='Allows the user to fold the element via touch or mouse.', verbose_name='touch enabled')),
                ('touch_sensitivity', models.DecimalField(decimal_places=2, default=0.25, help_text='Coefficient of touch/drag action’s distance delta. Higher numbers cause more movement.', max_digits=4, verbose_name='touch sensitivity')),
                ('touch_start_callback', models.CharField(blank=True, help_text='Invoked with starting coordinate as first argument.', max_length=255, verbose_name='touchstart callback')),
                ('touch_move_callback', models.CharField(blank=True, help_text='Invoked with the folded angle.', max_length=255, verbose_name='touchmove callback')),
                ('touch_end_callback', models.CharField(blank=True, help_text='Invoked with ending point.', max_length=255, verbose_name='touchend callback')),
            ],
            options={
                'verbose_name': 'Oridomi container',
                'verbose_name_plural': 'Oridomi containers',
            },
            bases=('cms.cmsplugin',),
        ),
    ]
